writers_rollingfilewriter.go

```diff
@@ -138,7 +140,7 @@ var compressionTypes = map[rollingArchiveType]compressionType{
 			if err != nil {
 				return err
 			}
-			return createGzip(archiveName, tar)
+			return createGzip(archiveName, bytes.NewReader(tar))
```

Comment:
I figured out how to hide the header details for the copy case (archived with the same format) and expose it reasonably for the add case (read from a file). I'll likely pull it out into a separate package, but here's the basic idea:
```
type archiveCopier interface {
	Copy() error
}

type archiveWriter interface {
	WriteHeaderFor(fi *os.FileInfo) error
	io.Writer
}

type archiveCopyWriter interface {
	archiveCopier
	archiveWriter
}

type tarCopyWriter struct {
	tr *tar.Reader
	tarWriter
}

func (t tarCopyWriter) Copy() error {
	tr := tar.NewReader(r)

	for {
		hdr, err := tr.Next()
		switch err {
		case io.EOF:
			break
		default: // Handle error
			return nil, err
		case nil: // Proceed below
		}

		info := hdr.FileInfo()
		// Skip directories
		if info.IsDir() {
			continue
		}
		if err := tarWriter.writeHeader(hdr); err != nil {
			return err
		}
		if _, err := io.Copy(tarWriter.tw, tr); err != nil {
			return err
		}
	}
	return files, nil
}

type tarWriter struct {
	tw *tar.Writer
	io.Writer
}

func (t tarWriter) WriteHeaderFor(fi *os.FileInfo) error {
	hdr, err := tar.FileInfoHeader(fi)
	if err != nil {
		return err
	}
	return t.writeHeader(hdr)
}

func (t tarWriter) writeHeader(hdr *tar.Header) error {
	return t.tw.WriteHeader(hdr)
}
```